### DISCUSSION TITLE
blog/unleash-the-power-of-fragments-with-graphql-codegen

### DISCUSSION THREAD

**Discussion by giscus** [the-guild-org/the-guild-website#918] [Docs Discussions]
# blog/unleash-the-power-of-fragments-with-graphql-codegen

The most important parts of Relay are the concepts of building and scaling applications, let's show how you can use these patterns in your existing project, no matter what client library you are currently using.

https://the-guild.dev/blog/unleash-the-power-of-fragments-with-graphql-codegen

**Comment by sfarthin** [comment]
I cannot for the life of me figure out how I can do the `enumsAsTypes: true` typescript option with the client preset. How can I do this?

**Reply by n1ru4l** [reply]
See https://github.com/dotansimha/graphql-code-generator/issues/8562

**Comment by khzouroussama** [comment]
i think there is a mistake there `UserList_QueryFragment` at the top-level component should actually be `FriendList_QueryFragment` right?

**Reply by n1ru4l** [reply]
Yes you are right, we fixed it 🎉

**Comment by sfarthin** [comment]
How do we get syntax highlighting for this API with https://marketplace.visualstudio.com/items?itemName=GraphQL.vscode-graphql-syntax&ssr=false#overview ? ```graphql(`query myQuery { ... }`)``` doesn't have syntax highlighting , but ```gql`query myQuery { ... }` ``` does.

**Reply by sfarthin** [reply]
I think `kumar-harsh.graphql-for-vscode@1.15.3` was breaking highlighting. Disregard.

**Comment by sfarthin** [comment]
I cannot figure out how to get your [eslint rules](https://the-guild.dev/graphql/eslint/docs/getting-started) to work with this setup. Is there a standalone project example I can refer to?

**Reply by dimaMachina** [reply]
graphql-eslint contain a lot of examples https://github.com/B2o5T/graphql-eslint/tree/master/examples

**Reply by sfarthin** [reply]
I've been looking through those, but I cannot seem to get them to work in my project. Here is a scaled down example I cannot seem to get to work: https://codesandbox.io/p/sandbox/confident-kilby-wj0ssg .

**Comment by ZakKa89** [comment]
This has complicated things a lot imo, typing is a hell now once you use fragments. Each time I have to use the FragmentType helper or useFragment or both. Tried to type a state variable with this, wasn't even able to figure it out. In previous versions this was much simpler by just importing the fragment type and you'd be done.

**Reply by n1ru4l** [reply]
We are using this at scale while building GraphQL Hive, and so far it has been a blessing and brought a lot of clarity to the data flow within our app.

Having that said, you can opt out if you don't like this and don't benefit from it. https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#how-to-disable-fragment-masking

Can you elaborate why you are trying to store the GraphQL result in a state variable?

**Reply by ZakKa89** [reply]
> We are using this at scale while building GraphQL Hive, and so far it has been a blessing and brought a lot of clarity to the data flow within our app.
> 
> Having that said, you can opt out if you don't like this and don't benefit from it. https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#how-to-disable-fragment-masking
> 
> Can you elaborate why you are trying to store the GraphQL result in a state variable?

Thank you for your reaction. 

Sure I can elaborate,  but isn't it a very typical use case to copy data into state? 

I have a use case where I'm fetching products, and the user has to select a product. I need to show the selected product in the UI to the user. (I could of course also only grab the id, and find the correct one based on that). 

See my question, and more feedback here regarding the use case: 
https://github.com/dotansimha/graphql-code-generator/discussions/9234

But honestly this is just one example of why I find it cumbersome to use, so I hope you take a look at the other points of my feedback as well. I opted out for fragment masking just over an hour ago and It's a blessing so far. 

I also read the relay docs to understand the principle behind it better and I just don't see the use case for my situtation:

https://relay.dev/docs/principles-and-architecture/thinking-in-relay/

If I understand correctly, one of the big reasons is to prevent that mutating data in one component can have an effect on another view. That all sounds fair but I don't need that at the moment and it's just a very big trade off to add all this code.

**Reply by n1ru4l** [reply]
> I could of course also only grab the id, and find the correct one based on that

This is the way to go.

**Reply by ZakKa89** [reply]
Sure in this particular use case that would solve it, but there might be other use cases where you want to put data in state... Would love to know if it's possible to type it with fragment masking on, but that's what the topic is for. 
 
Unless you're saying never ever do that and write your code in such a functional manner that you never need to put your data fragments into state. But can't imagine it being that uncommon. **Edit:** One example I could think of are the items inside a shopping cart: multiple products from a ProductFragment.

**Reply by max-ch9i** [reply]
You can use `DocumentType` to get the return type of `useFragment`.
```javascript
type Product = DocumentType<typeof ProductFragment>
```

I also responded to your discussion.

**Reply by ZakKa89** [reply]
Awesome thank you!

**Comment by RMHonor** [comment]
I wonder if with error paths in GQL responses it would be possible for individual fragment hooks to give an error, for more granular error handling.

**Reply by n1ru4l** [reply]
In terms of expected errors, we recommend modeling them as part of your GraphQL schema. See https://blog.logrocket.com/handling-graphql-errors-like-a-champ-with-unions-and-interfaces/

**Comment by ericbf** [comment]
Does `useFragment` have to follow the rules of hooks? Or rather, is there a way to use it in a non-hook context (for example, in a loop or function).

**Reply by dotansimha** [reply]
I think it does not have to follow it, as it's a simple extraction function that also infers types. You can see it's code here: https://github.com/dotansimha/graphql-code-generator/blob/72d1c1a22be239008daf888adf933c342236155c/examples/vue/urql/src/gql/fragment-masking.ts#L42

**Reply by ericbf** [reply]
Ok, cool! Looking at the implementation, it really is just a typecasting. I can rename the import to avoid the rules of hooks lint rule from flagging it safely then. Thanks!

**Comment by masterchef** [comment]
Thanks for the article! Can you explain the best way to test such components? Currently, I have to put together the input data from a set of fragments which is a little annoying. It would be great if I could define a my mock data as a single object and pass that in instead. For example:

// Currently only this works
```ts
const basicUser: BasicUserFragment = {
  id: 'test',
  ldap: 'ldap',
  firstName: 'First',
  fullName: 'First Last',
  __typename: 'UserResource',
};

const metricViewMock: MetricBlockFragment = {
  id: 'test',
  name: 'Metric Name',
  userMetricQuery: {
    ldap: 'ldap',
  },
  ownerUser: basicUser,
};
```

But it would be easier to just do this especially if there are a lot of nested fragments.

```ts
// One mock data object
const metricViewMock = {
  id: 'test',
  name: 'Metric Name',
  userMetricQuery: {
    ldap: 'ldap',
  },
  ownerUser: {
    id: 'test',
    ldap: 'ldap',
    firstName: 'First',
    fullName: 'First Last',
    __typename: 'UserResource',
  },
};
```

**Comment by satvikpendem** [comment]
Does this work with Flutter too?

**Comment by 2coo** [comment]
How to refetch specific fragment?
For example: in `FriendList` component, how to refetch `FriendList_QueryFragment` when user click refresh button?

**Reply by taprile314** [reply]
I also want to know, is there a way of only refetch fragment data or do I need to send up the refetch to my top-level component?

**Comment by taprile314** [comment]
Do this pattern pushes the developer to apply prop drilling?

**Comment by Gyv4ik** [comment]
Also interesting how to access data in deeply nested components using such approach.

**Reply by adriangalilea** [reply]
Same problem, there are several issues and PR's but all of them are being ignored.
https://github.com/dotansimha/graphql-code-generator/issues/9075
https://github.com/dotansimha/graphql-code-generator/pull/9380
https://github.com/dotansimha/graphql-code-generator/pull/9708

**Comment by thsyn** [comment]
I have a problem with Typescript (and apollo client) and don't know how to solve it.
I can't pass the query result to the child component without getting a type error.

The child component expects the type `FragmentType<typeof FRAGMENT>`

In the end, everything works, but there is a type error because the result of the apollo hook does not correspond to the FragmentType.
From my point of view, everything is implemented as in the example and I cannot find any other way to do it.



Root Component

```
'use client';

import { useSuspenseQuery } from '@apollo/client';

import { gql } from '#/types/__generated/gql';

import { TicketDetailsHeader } from './ticket-details-header';

const TicketQuery = gql(/* GraphQL */ `
  query Ticket($ticketNumber: Int!) {
    ticket(ticketNumber: $ticketNumber) {
      ...TicketHeaderFragment
    }
  }
`);

interface TicketDetailsProps {
  ticketNumber: number;
}

export function TicketDetails({ ticketNumber }: TicketDetailsProps) {
  const query = useSuspenseQuery(TicketQuery, { variables: { ticketNumber } });

  return (
    <>
      <TicketDetailsHeader query={query.data.ticket} />
    </>
  );
}
```

Child component
```
import { Typography } from '@mui/material';

import { formatTicketNumber } from '#/features/tickets/shared/format-ticket-number';
import { type FragmentType, gql, useFragment } from '#/types/__generated/gql';

const TicketHeaderFragment = gql(/* GraphQL */ `
  fragment TicketHeaderFragment on Ticket {
    id
    ticketNumber
    title
  }
`);

interface TicketDetailsHeaderProps {
  ticket: FragmentType<typeof TicketHeaderFragment>;
}

export function TicketDetailsHeader(props: TicketDetailsHeaderProps) {
  const ticket = useFragment(TicketHeaderFragment, props.ticket);
  const formattedTicketNumber = formatTicketNumber(ticket.ticketNumber);

  return (
    <>
      <Typography variant="h1">
        #{formattedTicketNumber}: {ticket.title}
      </Typography>
    </>
  );
}
```

**Reply by thsyn** [reply]
There is a typo

<TicketDetailsHeader query={query.data.ticket} />

should be
<TicketDetailsHeader ticket={query.data.ticket} />

**Reply by thsyn** [reply]
now.. ;)

`<TicketDetailsHeader query={query.data.ticket} />`
=> `<TicketDetailsHeader ticket={query.data.ticket} />`
